src/render/PointArray.cpp

```diff
@@ -386,14 +386,83 @@ bool PointArray::loadFile(QString fileName, size_t maxPointCount)
     {
         g_logger.debug("Reordering field %d: %s", i, m_fields[i]);
         reorder(m_fields[i], inds.get(), m_npoints);
-        emit loadProgress(int(100*(i+1)/m_fields.size()));
+        emit loadProgress(int(100*(i+1)/(m_fields.size()+1))); // denominator +1 for permutation reorder below
     }
     m_P = (V3f*)m_fields[m_positionFieldIdx].as<float>();
+    // The index we want to store is the reverse permutation of the index above
+    m_inds = std::unique_ptr<size_t[]>(new size_t[m_npoints]);
+    for (size_t i = 0; i < m_npoints; ++i) m_inds[inds[i]] = i;
+    emit loadProgress(int(100));
 
     return true;
 }
 
 
+void PointArray::mutate(std::shared_ptr<GeometryMutator> mutator)
+{
+    // Now we need to find the matching columns
+    auto npoints = mutator->pointCount();
+    const std::vector<GeomField>& mut_fields = mutator->fields();
+    auto mut_idx = mutator->index();
+
+    if (m_npoints > 4294967296)
+    {
+        g_logger.error("Mutation with more than 2^32 points is not supported");
+        return;
+    }
+
+    // Check index is valid
+    for (size_t j = 0; j < npoints; ++j)
+    {
+        if (mut_idx[j] < 0 || mut_idx[j] >= m_npoints)
+        {
+            g_logger.error("Index out of bounds - got %d (should be between zero and %d)", mut_idx[j], m_npoints-1);
+            return;
+        }
+    }
+
+    // Currently doesn't matter if a field isn't found
+    // Convenient way to ignore "index" but is not robust
+    for (size_t mut_field_idx = 0; mut_field_idx < mut_fields.size(); ++mut_field_idx)
+    {
+        if (mut_fields[mut_field_idx].name == "index")
+            continue;
+
+        bool found = false;
+
+        for (size_t field_idx = 0; field_idx < m_fields.size(); ++field_idx)
+        {
+
+            if (m_fields[field_idx].name == mut_fields[mut_field_idx].name)
```

Comment:
For breaking multiple loops?

Would occasionally be helpful, though I usually find it's a sign that a helper function is in order (`return` can be used just fine from deeply nested loops)